### DISCUSSION TITLE
Exponential back-off logging

### DISCUSSION THREAD

**Discussion by knoepfel** [gabime/spdlog#2559] [Q&A]
I help develop a [computing framework](https://github.com/art-framework-suite) that particle-physics experiments use to process their data.  The framework contains a logging facility that I'd love to replace with spdlog.  However, one of the features we provide is an exponential back-off of messages logged to a particular sink.

For example, a user could write something like:

```cpp
some_logger::set_backoff_factor(2);
for (std::size_t i = 1; i != 100; ++i) {
  some_logger::info("Processing {}", i);
}
```
and the output would be something like:
```console
Processing 1
Processing 2
Processing 4
Processing 8
Processing 16
Processing 32
Processing 64
```
Does spdlog have such a back-off facility (perhaps through a custom logger)?  Or would this have to be a new feature?

**Comment by tt4g** [comment]
Backtrace of logger is available

> ```cpp
> // Debug messages can be stored in a ring buffer instead of being logged immediately.
> // This is useful in order to display debug logs only when really needed (e.g. when error happens).
> // When needed, call dump_backtrace() to see them.
>
> spdlog::enable_backtrace(32); // Store the latest 32 messages in a buffer. Older messages will be dropped.
> // or my_logger->enable_backtrace(32)..
> for(int i = 0; i < 100; i++)
> {
>   spdlog::debug("Backtrace message {}", i); // not logged yet..
> }
> // e.g. if some error happened:
> spdlog::dump_backtrace(); // log them now! show the last 32 messages
>
> // or my_logger->dump_backtrace(32)..
> ````
>
> https://github.com/gabime/spdlog/tree/v1.11.0#backtrace-support

There is no feature to back off the logs of a specific sink, so it must be implemented custom sink.

See the Wiki for a custom sink implementation: https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink
See `spdlog::logger::backtrace()` for how to implement back off.

https://github.com/gabime/spdlog/blob/ad0e89cbfb4d0c1ce4d097e134eb7be67baebb36/include/spdlog/logger-inl.h#L107-L122

https://github.com/gabime/spdlog/blob/ad0e89cbfb4d0c1ce4d097e134eb7be67baebb36/include/spdlog/logger-inl.h#L210-L219

**Comment by tt4g** [comment]
`spdlog::sinks::ringbuffer_sink` is available, although it may be slightly different from the required back-off.

https://github.com/gabime/spdlog/blob/v1.11.0/include/spdlog/sinks/ringbuffer_sink.h
